### PR TITLE
Port REACT_MODULE() from react-native-windows to react-native-macos (1/4)

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

This is the first change for porting REACT_MODULE() from react-native-windows
It enables user to write a turbo module in C++ like this
```c++
REACT_MODULE(MyModule)
struct MyModule
{
    REACT_METHOD(MyMethod)
    void MyMethod(string a, int b, ReactPromise<double> result) noexcept
    {
        result.Reject("Not implemented");
    }
};
```

- This change moves some files from react-native-windows, they defines `REACT_MODULE` with a proper reflection system
- All copied files will eventually be downloaded from packages, the duplication is just a temporary solution for developing
- This change does not actually execute a turbo module written using `REACT_MODULE`, this will be done in future pull requests

## For Code Review

The original code is deeply coupled with WinRT,
I have no choice but to implement involved WinRT API in macOS,
here is two files that do this job

- **RNTester/RNTester-macOS/TurboModuleCxx/winrt/Windows.Foundation.h**
- **RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h**

For other files in **RNTester/RNTester-macOS/TurboModuleCxx**,
they are copied from react-native-windows,
which mean I am not going to fix issues about these files in this pull request
(because such changes are expected to be submitted to react-native-windows),
except for code in `#ifdef __APPLE__`.

As what is mentioned above,
eventually these files will be downloaded from a package,
so after a few pull requests,
they will be deleted from this repo.

Having duplicated files here is a temporary solution for development convenience.

## About Future Pull Requests

More works are needed to make this scenario completes.
In the next pull request,
I will make turbo modules written using `REACT_MODULE` executable.

## Changelog

[macOS] [Fixed] - Port REACT_MODULE() from react-native-windows to react-native-macos (1/4)

## Test Plan

- The same code has been used in react-native-windows for quite a long time
- Since the execution part is not completed, so no test is applicable here

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-macos/pull/595)

### CODE REVIEW

File: RNTester/RNTesterPods.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1852,6 +1929,31 @@
...
+					"\"${PODS_ROOT}/Headers/Public/React-jsi\"",
+					"\"${PODS_ROOT}/Headers/Public/React-jsiexecutor\"",
+					"\"${PODS_ROOT}/Headers/Public/React-jsinspector\"",
+					"\"${PODS_ROOT}/Headers/Public/ReactCommon\"",
+					"\"${PODS_ROOT}/Headers/Public/Yoga\"",
+					"\"${PODS_ROOT}/Headers/Public/glog\"",
+					"\"${PODS_ROOT}/Headers/Public/hermes\"",
+					"\"${PODS_ROOT}/Headers/Public/libevent\"",
+					"\"$(PODS_ROOT)/boost-for-react-native\"",
+					"\"$(PODS_ROOT)/Headers/Private/React-Core\"",
```
Comment (alloy) [inline RNTester/RNTesterPods.xcodeproj/project.pbxproj:1954]: Does your code really need _all_ of the above search paths or is this a merge issue?

**Comment by alloy** [comment]
I have a bit of hard time judging what’s what. Does this change really add any macOS specific code yet or is it just the foundation that you will apply the needed macOS changes on top of?

File: RNTester/RNTesterPods.xcodeproj/project.pbxproj (hunk 1 above)
Comment (ZihanChen-MSFT) [reply RNTester/RNTesterPods.xcodeproj/project.pbxproj:1954]: Actually I just add the last line, but seems xcode think that RNTester-macOS should inherit from what is in RNTester, so everything is copied to RNTester-macOS

**Comment by ZihanChen-MSFT** [comment]
> I have a bit of hard time judging what’s what. Does this change really add any macOS specific code yet or is it just the foundation that you will apply the needed macOS changes on top of?
> 
> The [CI failure](https://dev.azure.com/ms/react-native/_build/results?buildId=108819&view=logs&j=628dfcdf-afa5-5702-e4da-dfa310a93062&t=65c699ef-104f-5f6e-4efa-350cb1662389) seems legit btw.

Actually there are only 3 files that added to the project:

- RNTester/RNTester-macOS/TurboModuleCxx/winrt/Windows.Foundation.h
  - declares/implements all involved C++/WinRT APIs
- RNTester/RNTester-macOS/TurboModuleCxx/winrt/Microsoft.ReactNative.h
  -  declares/implements all involved COM object that is defined in `react-native-windows`
- ScreenshotManager.mm
  - Added a C++ turbo module to check if the code compiles

File: RNTester/RNTesterPods.xcodeproj/project.pbxproj (hunk 1 above)
Comment (alloy) [reply RNTester/RNTesterPods.xcodeproj/project.pbxproj:1954]: Ah I see. Can you remove them all and instead add `$(inherited)`?

**Review by alloy — APPROVED** [review]
At some point this will need to be installed through CocoaPods, but we can defer that until a later stage 👍

**Comment by alloy** [comment]
Looks like your branch might be too out of date. I expect the final CI failure to go away if you first update it.

File: RNTester/RNTesterPods.xcodeproj/project.pbxproj (hunk 1 above)
Comment (ZihanChen-MSFT) [reply RNTester/RNTesterPods.xcodeproj/project.pbxproj:1954]: I can do that when I do the package importing, could be the last pull request. I think I will remove folders in RNTester-macOS and add a line in RNTester.

**Pull request merged**